File: agent/src/repos/recent_files_store.ts
// Path: agent/src/repos/recent_files_store.ts
// Description: Persistence layer for recent files with debounced atomic writes

import * as fs from "node:fs";
import * as fsPromises from "node:fs/promises";
import * as path from "node:path";
import * as crypto from "node:crypto";
import type { FileEntry } from "../../../app/src/shared/protocol.js";
import { logger } from "../util/logger.js";

const PERSIST_DEBOUNCE_MS = 500;
const SCHEMA_VERSION = 1;

interface PersistedRecentFiles {
  version: number;
  repoId: string;
  repoRoot: string;
  updatedAtIso: string;
  entries: FileEntry[];
}

export interface RecentFilesStore {
  /** Sync load on startup; returns [] if missing/corrupt */
  load(repoId: string, repoRoot: string): FileEntry[];
  /** Schedule debounced save */
  scheduleSave(repoId: string, repoRoot: string, entries: FileEntry[]): void;
  /** Cancel pending save for a repo */
  cancelPending(repoId: string): void;
  /** Flush all pending writes immediately (for shutdown) */
  flush(): Promise<void>;
}

function getFilePath(stateDir: string, repoId: string): string {
  return path.join(stateDir, "recent_files", `${repoId}.json`);
}

function isValidSchema(data: unknown): data is PersistedRecentFiles {
  if (typeof data !== "object" || data === null) return false;
  const obj = data as Record<string, unknown>;
  return (
    typeof obj["version"] === "number" &&
    typeof obj["repoId"] === "string" &&
    typeof obj["repoRoot"] === "string" &&
    typeof obj["updatedAtIso"] === "string" &&
    Array.isArray(obj["entries"])
  );
}

export function createRecentFilesStore(stateDir: string): RecentFilesStore {
  const pendingWrites = new Map<string, NodeJS.Timeout>();
  const pendingData = new Map<string, { repoRoot: string; entries: FileEntry[] }>();

  function load(repoId: string, repoRoot: string): FileEntry[] {
    const filePath = getFilePath(stateDir, repoId);
    try {
      if (!fs.existsSync(filePath)) {
        logger.debug("No persisted recent files found", { repoId, filePath });
        return [];
      }

      const content = fs.readFileSync(filePath, "utf-8");
      const data: unknown = JSON.parse(content);

      if (!isValidSchema(data)) {
        logger.warn("Invalid recent files schema", { repoId, filePath });
        return [];
      }

      if (data.version !== SCHEMA_VERSION) {
        logger.warn("Unsupported recent files version", {
          repoId,
          found: data.version,
          expected: SCHEMA_VERSION,
        });
        return [];
      }

      if (data.repoRoot !== repoRoot) {
        logger.info("Repo root changed, keeping entries", {
          repoId,
          stored: data.repoRoot,
          current: repoRoot,
        });
      }

      logger.debug("Loaded persisted recent files", {
        repoId,
        count: data.entries.length,
      });
      return data.entries;
    } catch (err) {
      if (err instanceof SyntaxError) {
        logger.warn("Corrupt recent files JSON", { repoId, filePath });
      } else {
        logger.error("Failed to load recent files", {
          repoId,
          error: err instanceof Error ? err.message : String(err),
        });
      }
      return [];
    }
  }

  function scheduleSave(repoId: string, repoRoot: string, entries: FileEntry[]): void {
    // Store the latest data
    pendingData.set(repoId, { repoRoot, entries });

    // Cancel existing timeout
    const existing = pendingWrites.get(repoId);
    if (existing) {
      clearTimeout(existing);
    }

    // Schedule new debounced write
    const timeout = setTimeout(() => {
      pendingWrites.delete(repoId);
      void writeFile(repoId);
    }, PERSIST_DEBOUNCE_MS);

    pendingWrites.set(repoId, timeout);
  }

  function cancelPending(repoId: string): void {
    const existing = pendingWrites.get(repoId);
    if (existing) {
      clearTimeout(existing);
      pendingWrites.delete(repoId);
    }
    pendingData.delete(repoId);
  }

  async function writeFile(repoId: string): Promise<void> {
    const data = pendingData.get(repoId);
    if (!data) return;
    pendingData.delete(repoId);

    const filePath = getFilePath(stateDir, repoId);
    const dir = path.dirname(filePath);

    const payload: PersistedRecentFiles = {
      version: SCHEMA_VERSION,
      repoId,
      repoRoot: data.repoRoot,
      updatedAtIso: new Date().toISOString(),
      entries: data.entries,
    };

    try {
      await fsPromises.mkdir(dir, { recursive: true });

      // Atomic write: temp file + rename
      const tempPath = `${filePath}.${crypto.randomUUID()}.tmp`;
      await fsPromises.writeFile(tempPath, JSON.stringify(payload, null, 2), "utf-8");
      await fsPromises.rename(tempPath, filePath);

      logger.debug("Persisted recent files", { repoId, count: data.entries.length });
    } catch (err) {
      logger.error("Failed to persist recent files", {
        repoId,
        error: err instanceof Error ? err.message : String(err),
      });
    }
  }

  async function flush(): Promise<void> {
    // Cancel all pending timeouts
    for (const timeout of pendingWrites.values()) {
      clearTimeout(timeout);
    }
    pendingWrites.clear();

    // Write all pending data immediately
    const writes = Array.from(pendingData.keys()).map((repoId) => writeFile(repoId));
    await Promise.all(writes);
  }

  return {
    load,
    scheduleSave,
    cancelPending,
    flush,
  };
}

File: crates/im_agent/src/repos/recent_files_store.rs
// Path: crates/im_agent/src/repos/recent_files_store.rs
// Description: Persist recent files with debounced atomic writes

use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::time::Duration;

use serde::{Deserialize, Serialize};
use tokio::fs;
use tokio::sync::Mutex;
use tokio::time::sleep;

use crate::logging::Logger;
use crate::protocol::FileEntry;

const PERSIST_DEBOUNCE_MS: u64 = 500;
const SCHEMA_VERSION: u32 = 1;

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct PersistedRecentFiles {
    version: u32,
    repo_id: String,
    repo_root: String,
    updated_at_iso: String,
    entries: Vec<FileEntry>,
}

struct PendingWrite {
    repo_root: String,
    entries: Vec<FileEntry>,
    handle: Option<tokio::task::JoinHandle<()>>,
}

#[derive(Clone)]
pub struct RecentFilesStore {
    state_dir: PathBuf,
    logger: Logger,
    pending: std::sync::Arc<Mutex<HashMap<String, PendingWrite>>>,
}

impl RecentFilesStore {
    pub fn new(state_dir: PathBuf, logger: Logger) -> Self {
        Self {
            state_dir,
            logger,
            pending: std::sync::Arc::new(Mutex::new(HashMap::new())),
        }
    }

    pub async fn load(&self, repo_id: &str, repo_root: &str) -> Vec<FileEntry> {
        let file_path = self.file_path(repo_id);
        if fs::metadata(&file_path).await.is_err() {
            return Vec::new();
        }

        let content = match fs::read_to_string(&file_path).await {
            Ok(content) => content,
            Err(err) => {
                self.logger.warn(
                    "Failed to read recent files",
                    Some(serde_json::json!({"repoId": repo_id, "error": err.to_string()})),
                );
                return Vec::new();
            }
        };

        let data: PersistedRecentFiles = match serde_json::from_str(&content) {
            Ok(data) => data,
            Err(_) => {
                self.logger.warn(
                    "Corrupt recent files JSON",
                    Some(serde_json::json!({"repoId": repo_id})),
                );
                return Vec::new();
            }
        };

        if data.version != SCHEMA_VERSION {
            self.logger.warn(
                "Unsupported recent files schema",
                Some(serde_json::json!({"repoId": repo_id, "found": data.version, "expected": SCHEMA_VERSION})),
            );
            return Vec::new();
        }

        if data.repo_root != repo_root {
            self.logger.info(
                "Repo root changed for recent files",
                Some(serde_json::json!({"repoId": repo_id, "stored": data.repo_root, "current": repo_root})),
            );
        }

        data.entries
    }

    pub async fn schedule_save(&self, repo_id: String, repo_root: String, entries: Vec<FileEntry>) {
        let store = self.clone();
        let repo_id_clone = repo_id.clone();

        let mut pending = self.pending.lock().await;
        if let Some(existing) = pending.get_mut(&repo_id) {
            if let Some(handle) = existing.handle.take() {
                handle.abort();
            }
            existing.repo_root = repo_root;
            existing.entries = entries;
        } else {
            pending.insert(
                repo_id.clone(),
                PendingWrite {
                    repo_root,
                    entries,
                    handle: None,
                },
            );
        }

        let handle = tokio::spawn(async move {
            sleep(Duration::from_millis(PERSIST_DEBOUNCE_MS)).await;
            store.write_pending(&repo_id_clone).await;
        });

        if let Some(existing) = pending.get_mut(&repo_id) {
            existing.handle = Some(handle);
        }
    }

    pub async fn flush_repo(&self, repo_id: &str) {
        let pending_entry = { self.pending.lock().await.remove(repo_id) };

        if let Some(mut entry) = pending_entry {
            if let Some(handle) = entry.handle.take() {
                handle.abort();
            }
            self.write_now(repo_id.to_string(), entry.repo_root, entry.entries)
                .await;
        }
    }

    pub async fn flush_all(&self) {
        let entries = {
            let mut pending = self.pending.lock().await;
            pending
                .drain()
                .map(|(key, value)| (key, value))
                .collect::<Vec<_>>()
        };

        for (repo_id, entry) in entries {
            let mut entry = entry;
            if let Some(handle) = entry.handle.take() {
                handle.abort();
            }
            self.write_now(repo_id, entry.repo_root, entry.entries)
                .await;
        }
    }

    async fn write_pending(&self, repo_id: &str) {
        let pending_entry = { self.pending.lock().await.remove(repo_id) };
        if let Some(entry) = pending_entry {
            self.write_now(repo_id.to_string(), entry.repo_root, entry.entries)
                .await;
        }
    }

    async fn write_now(&self, repo_id: String, repo_root: String, entries: Vec<FileEntry>) {
        let file_path = self.file_path(&repo_id);
        let dir = file_path.parent().unwrap_or_else(|| Path::new("."));

        let payload = PersistedRecentFiles {
            version: SCHEMA_VERSION,
            repo_id,
            repo_root,
            updated_at_iso: chrono::Utc::now().to_rfc3339(),
            entries,
        };

        if let Err(err) = fs::create_dir_all(dir).await {
            self.logger.error(
                "Failed to create recent files dir",
                Some(serde_json::json!({"error": err.to_string()})),
            );
            return;
        }

        let temp_path = file_path.with_extension(format!(
            "tmp-{}",
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|duration| duration.as_nanos())
                .unwrap_or(0)
        ));
        let content = match serde_json::to_string_pretty(&payload) {
            Ok(content) => content,
            Err(err) => {
                self.logger.error(
                    "Failed to serialize recent files",
                    Some(serde_json::json!({"error": err.to_string()})),
                );
                return;
            }
        };

        if let Err(err) = fs::write(&temp_path, content).await {
            self.logger.error(
                "Failed to write recent files",
                Some(serde_json::json!({"error": err.to_string()})),
            );
            let _ = fs::remove_file(&temp_path).await;
            return;
        }

        if let Err(err) = fs::rename(&temp_path, &file_path).await {
            self.logger.error(
                "Failed to finalize recent files",
                Some(serde_json::json!({"error": err.to_string()})),
            );
            let _ = fs::remove_file(&temp_path).await;
        }
    }

    fn file_path(&self, repo_id: &str) -> PathBuf {
        self.state_dir
            .join("recent_files")
            .join(format!("{repo_id}.json"))
    }
}
